include/oneapi/tbb/detail/_flow_graph_impl.h

```diff
@@ -349,30 +346,50 @@ class graph : no_copy, public graph_proxy {
 
     graph_task_priority_queue_t my_priority_queue;
 
+    d1::wait_context_vertex& get_wait_context_vertex() { return my_wait_context_vertex; }
+
     friend void activate_graph(graph& g);
     friend void deactivate_graph(graph& g);
     friend bool is_graph_active(graph& g);
+    friend bool is_thread_in_graph_arena(graph& g);
     friend graph_task* prioritize_task(graph& g, graph_task& arena_task);
     friend void spawn_in_graph_arena(graph& g, graph_task& arena_task);
     friend void enqueue_in_graph_arena(graph &g, graph_task& arena_task);
 
     friend class task_arena_base;
-
+    friend class graph_task;
 };  // class graph
 
 template<typename DerivedType>
-inline void graph_task::destruct_and_deallocate(const execution_data& ed) {
+inline void graph_task::destruct_and_deallocate(const d1::execution_data& ed) {
     auto allocator = my_allocator;
     // TODO: investigate if direct call of derived destructor gives any benefits.
     this->~graph_task();
     allocator.deallocate(static_cast<DerivedType*>(this), ed);
 }
 
 template<typename DerivedType>
-inline void graph_task::finalize(const execution_data& ed) {
-    graph& g = my_graph;
+inline void graph_task::finalize(const d1::execution_data& ed) {
+    d1::wait_tree_vertex_interface* reference_vertex = my_reference_vertex;
     destruct_and_deallocate<DerivedType>(ed);
-    g.release_wait();
+    reference_vertex->release();
+}
+
+inline graph_task::graph_task(graph& g, d1::small_object_allocator& allocator,
+                              node_priority_t node_priority)
+    : my_graph(g)
+    , priority(node_priority)
+    , my_allocator(allocator)
+{
+    // If the task is created by the thread outside the graph arena, the lifetime of the thread reference vertex
+    // may be shorter that the lifetime of the task, so thread reference vertex approach cannot be used
+    // and the task should be associated with the graph wait context itself
+    // TODO: consider how reference counting can be improved for such a use case. Most common example is the async_node
+    d1::wait_context_vertex* graph_wait_context_vertex = &my_graph.get_wait_context_vertex();
+    my_reference_vertex = is_thread_in_graph_arena(g) ? r1::get_thread_reference_vertex(graph_wait_context_vertex)
```

Comment:
I believe this can happen even without `async_node`. Consider the following case:
```cpp
std::thread t1([]() {
            tbb::task_arena& gta = get_graph_task_arena();
            gta.execute([]() {
                             auto& some_graph_node = get_graph_node();
                             some_graph_node.try_put(/*node message*/1); // spawn of task A
            });
});

t1.join();


// task A starts its execution
graph.wait_for_all();
```
Don't we support such cases anymore?

Reply:
In this case it is guaranteed that `t1` will either join the `gta` and execute the task if there is an available slot or send a task and it would be executed by another worker in `gta`. In each case, the thread executing the lambda and calling `try_put` is guaranteed to be in `gta` and hence has its reference node in the TLS (the code will choose the `true` branch). 
If `t1` would submit work in the graph without explicitly calling `task_arena::execute`, the `false` branch would be chosen and the task reference counting would work as before this PR by adding a reference to graph wait_context.